### PR TITLE
sum-of-multiples: handle cardinality > 3 correctly

### LINKED ISSUE #1367: sum-of-multiples: handle cardinailty > 3 correctly
It is possible to approach this exercises using an entirely arithmetic approach instead of a container type. In this approach the total of the [arithmetic series](http://mathworld.wolfram.com/ArithmeticSeries.html) of each factor must be compensated for overlaps using the [inclusion-exclusion principle](http://mathworld.wolfram.com/Inclusion-ExclusionPrinciple.html); the current body of tests handle only up to a cardinality of 3, but it is quite easy to get something passing a cardinality of 3 without realizing it will fail with cardinality of 4 or more. A test with 5 prime factors will demonstrate the ability of the solution to handle the problem generally.

### PR DESCRIPTION
**Pull request by yawpitch**
Closes #1367.

For some reason the diff is currently showing against the 1.2.0 version, but that's not what appears to be in master (1.3.0). My only change is the addition of the `multiples of 5 primes up to 100` test case.

### CODE REVIEW

File: exercises/sum-of-multiples/canonical-data.json [hunk 1]
```diff
@@ -118,6 +118,24 @@
...
+      "description": "the only multiple of 0 is 0",
+      "property": "sum",
+      "input": {
+        "factors": [0],
+        "limit": 1
+      },
+      "expected": 0
+    },
+    {
+      "description": "multiples of 5 primes up to 100",
```
[the remark this answers is not in the document]
Comment (yawpitch) [reply exercises/sum-of-multiples/canonical-data.json]: Working on fixing the diff; seems I had a fork of this repo hanging around I didn't remember ever making. Sigh. Will fix commit message as well.

Personally agree with the lack of the need for `up to 100` but I put it in there to remain consistent with all the other descriptions; the limit isn't really the important part of any of those cases either.

[the remark this answers is not in the document]
Comment (petertseng) [reply exercises/sum-of-multiples/canonical-data.json]: My *personal* preference is to favour making sure that new additions focus on what's important, and a preference for this over consistency with other cases. It is true that it is a problem for other cases, but I would only suggest adding to the problem if the benefit to doing so overcomes the addition. I was thinking that the benefit of the consistency did not outweigh.

[the remark this answers is not in the document]
Comment (petertseng) [reply exercises/sum-of-multiples/canonical-data.json]: This is interesting because I agree that this case is most necessary for solutions using the inclusion-exclusion principle; solutions using filtering that get to this point likely can already handle this.

I don't have a strong opinion on whether the test description needs to be so specific as to the implementations that it targets. I'll let you make that choice and accept either outcome, though I suspect it will be useful if the description still mentions cardinality > 3

**Comment by petertseng** [comment]
For anyone who may be wondering whether anyone has checked: I have agreed that the expected answer should be 3917.

**Comment by petertseng** [comment]
Here's an interesting question...

Would a larger limit for this case make sense? I think it needs to be `>= 2 * 3 * 5 * 7 * 11` (they're all relatively prime) to test this, right? As it stands, this case doesn't check whether a solution would exclude multiples of `2 * 3 * 5 * 7` and include multiples of `2 * 3 * 5 * 7 * 11`

**Comment by petertseng** [comment]
I just added an implementation that is broken for cardinality > 3, but it doesn't fail any tests (including this newly proposed one) so I think the limit needs to be larger. Implementation below.

```ruby
  {
    name: 'inclusion-exclusion principle except broken for n > 3',
    should_fail: true,
    f: ->(i, _) {
      factors = i['factors']
      signed_factors = (1..factors.size).flat_map { |n| 
        # Even terms are subtracted, so 0 -> -1.
        # Odd terms are added, so 1 -> 1.
        sign = n > 3 ? 100000 : 2 * (n % 2) - 1 
        factors.combination(n).map { |n_factors|
          n_factors.reduce(1) { |a, b| a.lcm(b) } * sign
        }   
      }.freeze
      limit = i['limit']

      signed_factors.sum { |n| 
        # how many times does it fit?
        n_terms = n == 0 ? 0 : (limit - 1) / n.abs
        n * n_terms * (n_terms + 1) / 2 
      }   
    },  
  },  
```

**Comment by yawpitch** [comment]
Hmm, good point. The specific implementation I got from a student failed at precisely cardinality 4+ in all cases, including with a limit of 100, but that was because it simply didn't even try to compensate at all beyond 3. If I raise to 10000 it should catch that.

**Comment by yawpitch** [comment]
Thanks for the solid review.

File: exercises/sum-of-multiples/canonical-data.json [hunk 2]
```diff
@@ -127,6 +127,15 @@
...
+      "input": {
+        "factors": [2, 3, 5, 7, 11],
+        "limit": 10000
+      },
+      "expected": 39614537
     }
   ]
-}
\ No newline at end of file
+}
```
Comment (petertseng) [inline exercises/sum-of-multiples/canonical-data.json]: as a matter of course I recommend that commits don't include unrelated changes such as whitespace

I just got bit by this when making my PR to sum-of-multiples though since my editor automatically adds the trailing newline on save. Perhaps this repo is needing an automatic formatter so that I don't have to expend routine objections like this. something like `jq .` would do.

Comment (petertseng) [reply exercises/sum-of-multiples/canonical-data.json]: it doesn't matter to me whether this is changed, but I file the routine objection nevertheless

**Comment by petertseng** [comment]
Anyone who sees a problem and wishes to get a word in, I'd say you have about 48 hours to do so.

**Comment by petertseng** [comment]
Great stuff, thank you!
